src/hashes.rs

```diff
@@ -1,473 +1,313 @@
-            Self::Sha3_512 => Some(Box::new(Sha3_512)),
-            Self::Keccak224 => Some(Box::new(Keccak224)),
-            Self::Keccak256 => Some(Box::new(Keccak256)),
-            Self::Keccak384 => Some(Box::new(Keccak384)),
-            Self::Keccak512 => Some(Box::new(Keccak512)),
-            Self::Blake2b256 => Some(Box::new(Blake2b256)),
-            Self::Blake2b512 => Some(Box::new(Blake2b512)),
-            Self::Blake2s128 => Some(Box::new(Blake2s128)),
-            Self::Blake2s256 => Some(Box::new(Blake2s256)),
-        }
-    }
+    Blake2s256 => 0xb260,
 }
 
-/// The Identity hasher
-#[derive(Clone, Debug)]
-pub struct Identity;
+/// The Identity hasher.
+#[derive(Clone, Debug, Default)]
+pub struct Identity(Vec<u8>);
 impl MultihashDigest for Identity {
     fn code(&self) -> Code {
         Self::CODE
     }
     fn digest(&self, data: &[u8]) -> Multihash {
         Self::digest(data)
     }
-}
-impl Identity {
-    /// The code of Identity hasher, 0x00.
-    pub const CODE: Code = Code::Identity;
-    /// Hash some input and return the raw binary digest.
-    pub fn digest(data: &[u8]) -> Multihash {
-        if (data.len() as u64) >= u64::from(std::u32::MAX) {
-            panic!("Input data for identity hash is too large, it needs to be less the 2^32.")
+    fn input(&mut self, data: &[u8]) {
+        if ((self.0.len() as u64) + (data.len() as u64)) >= u64::from(std::u32::MAX) {
+            panic!("Input data for identity hash is too large, it needs to be less than 2^32.")
         }
-        wrap(Self::CODE, &data)
-    }
-}
-
-/// The Sha1 hasher.
-#[derive(Clone, Debug)]
-pub struct Sha1;
-impl MultihashDigest for Sha1 {
-    fn code(&self) -> Code {
-        Self::CODE
-    }
-    fn digest(&self, data: &[u8]) -> Multihash {
-        Self::digest(data)
-    }
-}
-impl Sha1 {
-    /// The code of Sha1 hasher, 0x11.
-    pub const CODE: Code = Code::Sha1;
-    /// Hash some input and return the sha1 digest.
-    pub fn digest(data: &[u8]) -> Multihash {
-        let digest = sha1::Sha1::digest(&data);
-        wrap(Self::CODE, &digest)
+        self.0.extend_from_slice(data)
     }
-}
-
-/// The Sha2-256 hasher.
-#[derive(Clone, Debug)]
-pub struct Sha2_256;
-impl MultihashDigest for Sha2_256 {
-    fn code(&self) -> Code {
-        Self::CODE
+    fn result(self) -> Multihash {
+        wrap(Self::CODE, &self.0)
     }
-    fn digest(&self, data: &[u8]) -> Multihash {
-        Self::digest(data)
+    fn result_reset(&mut self) -> Multihash {
+        let hash = wrap(Self::CODE, &self.0);
+        self.reset();
+        hash
     }
-}
-impl Sha2_256 {
-    /// The code of Sha2-256 hasher, 0x12.
-    pub const CODE: Code = Code::Sha2_256;
-    /// Hash some input and return the sha2-256 digest.
-    pub fn digest(data: &[u8]) -> Multihash {
-        let digest = sha2::Sha256::digest(&data);
-        wrap(Self::CODE, &digest)
+    fn reset(&mut self) {
+        self.0 = Vec::new();
```

Comment:
Should this be `self.0.clear()` instead? I'm unsure what exactly the `reset()` semantics are supposed to be.

Reply:
I think that's better, as the method docs only detail that it needs to "reset the hasher to its initial state": [docs](https://docs.rs/digest/0.8.1/digest/trait.Digest.html#tymethod.reset). I'll update when I get the chance.

Reply:
On second thought, the docs state that `clear()` "has no effect on the allocated capacity of the vector," which may or may not be desired behaviour as the `Vec` wouldn't literally be in its initial state...

Reply:
I would say it would be in it's initial state, it's just not the same state.

I think the point of `reset()` is that it is an operation that is cheaper than recreating it. Hence `clear()` would make sense. You could always create a new object if you want to start from a clean state.

Reply:
It's changed in the latest updates.